File: src/test/java/com/example/the_machine/operator/OperatorResultTest.java
package com.example.the_machine.operator;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.example.the_machine.operator.result.OperatorResult;
import com.example.the_machine.operator.result.OperatorResultCode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class OperatorResultTest {

  @Test
  void testSuccessFactory() {
    Map<String, Object> inputs = Map.of("param1", "value1");
    Map<String, Object> outputs = Map.of("result", "success");

    final var result = OperatorResult.success("TestOp", "1.0", inputs, outputs);

    assertTrue(result.isOk());
    assertEquals(OperatorResultCode.SUCCESS, result.getCode());
    assertEquals("Operation completed successfully", result.getMessage());
    assertEquals("TestOp", result.getOperatorName());
    assertEquals("1.0", result.getOperatorVersion());
    assertEquals(inputs, result.getNormalizedInputs());
    assertEquals(outputs, result.getOutputs());
    assertEquals(1, result.getAttempts());
  }

  @Test
  void testSuccessFactoryWithCustomMessage() {
    Map<String, Object> inputs = Map.of("param1", "value1");
    Map<String, Object> outputs = Map.of("result", "success");

    final var result = OperatorResult.success("TestOp", "1.0", "Custom success", inputs, outputs);

    assertTrue(result.isOk());
    assertEquals(OperatorResultCode.SUCCESS, result.getCode());
    assertEquals("Custom success", result.getMessage());
    assertEquals("TestOp", result.getOperatorName());
    assertEquals("1.0", result.getOperatorVersion());
  }

  @Test
  void testFailureFactory() {
    Map<String, Object> inputs = Map.of("param1", "value1");

    final var result = OperatorResult.failure("TestOp", "1.0", OperatorResultCode.VALIDATION_ERROR,
        "Parameter validation failed", inputs);

    assertFalse(result.isOk());
    assertEquals(OperatorResultCode.VALIDATION_ERROR, result.getCode());
    assertEquals("Parameter validation failed", result.getMessage());
    assertEquals("TestOp", result.getOperatorName());
    assertEquals("1.0", result.getOperatorVersion());
    assertEquals(inputs, result.getNormalizedInputs());
    assertTrue(result.getOutputs().isEmpty());
  }

  @Test
  void testValidationFailureFactory() {
    final var result = OperatorResult.validationFailure("TestOp", "1.0",
        "Required parameter 'name' is missing");

    assertFalse(result.isOk());
    assertEquals(OperatorResultCode.VALIDATION_ERROR, result.getCode());
    assertEquals("Required parameter 'name' is missing", result.getMessage());
    assertEquals("TestOp", result.getOperatorName());
    assertEquals("1.0", result.getOperatorVersion());
  }

  @Test
  void testWithAttempts() {
    final var original = OperatorResult.success("TestOp", "1.0", Map.of(), Map.of());
    final var updated = original.withAttempts(3);

    assertEquals(1, original.getAttempts());
    assertEquals(3, updated.getAttempts());

    // Ensure other fields are preserved
    assertEquals(original.isOk(), updated.isOk());
    assertEquals(original.getCode(), updated.getCode());
    assertEquals(original.getOperatorName(), updated.getOperatorName());
  }

  @Test
  void testWithMetrics() {
    final var original = OperatorResult.success("TestOp", "1.0", Map.of(), Map.of());
    Map<String, Object> additionalMetrics = Map.of("latency", 150L, "retries", 2);
    final var updated = original.withMetrics(additionalMetrics);

    assertTrue(original.getMetrics().isEmpty());
    assertEquals(150L, updated.getMetrics().get("latency"));
    assertEquals(2, updated.getMetrics().get("retries"));

    // Test merging with existing metrics
    Map<String, Object> existingMetrics = Map.of("cpu", 0.8);
    final var withExisting = OperatorResult.success("TestOp", "1.0", Map.of(), Map.of())
        .withMetrics(existingMetrics);

    final var merged = withExisting.withMetrics(additionalMetrics);
    assertEquals(0.8, merged.getMetrics().get("cpu"));
    assertEquals(150L, merged.getMetrics().get("latency"));
    assertEquals(2, merged.getMetrics().get("retries"));
  }

  @Test
  void testWithRemediationPatches() {
    final var original = OperatorResult.success("TestOp", "1.0", Map.of(), Map.of());
    final var objectMapper = new ObjectMapper();
    final var patch = objectMapper.createObjectNode().put("op", "replace").put("path", "/param1")
        .put("value", "newValue");
    final var patches = List.<com.fasterxml.jackson.databind.JsonNode>of(patch);

    final var updated = original.withRemediationPatches(patches);

    assertTrue(original.getRemediationPatches().isEmpty());
    assertEquals(1, updated.getRemediationPatches().size());
    assertEquals("replace", updated.getRemediationPatches().getFirst().get("op").asText());
  }

  @Test
  void testToString() {
    final var result = OperatorResult.success("TestOperator", "2.1", Map.of(), Map.of());
    final var toString = result.toString();

    assertTrue(toString.contains("ok=true"));
    assertTrue(toString.contains("code='SUCCESS'"));
    assertTrue(toString.contains("operator='TestOperator@2.1'"));
    assertTrue(toString.contains("attempts=1"));
    assertTrue(toString.contains("message='Operation completed successfully'"));
  }

  @Test
  void testNullHandling() {
    // Test with null inputs and outputs
    final var result = OperatorResult.success("TestOp", "1.0", null, null);

    assertNotNull(result.getNormalizedInputs());
    assertNotNull(result.getOutputs());
    assertTrue(result.getNormalizedInputs().isEmpty());
    assertTrue(result.getOutputs().isEmpty());
  }

  @Test
  void testCustomCodeViaFailureMethod() {
    final var result = OperatorResult.failure("CustomOp", "3.0",
        OperatorResultCode.EXECUTION_FAILED,
        "Custom operation completed", Map.of());

    assertFalse(result.isOk()); // failure method creates failed results
    assertEquals(OperatorResultCode.EXECUTION_FAILED, result.getCode());
    assertEquals("Custom operation completed", result.getMessage());
    assertEquals("CustomOp", result.getOperatorName());
    assertEquals("3.0", result.getOperatorVersion());
    assertEquals(1, result.getAttempts()); // default attempts
  }
}

File: src/test/kotlin/com/example/the_machine/operator/OperatorResultTest.kt
package com.example.the_machine.operator

import com.example.the_machine.common.KotlinSerializationConfig
import com.example.the_machine.operator.result.OperatorResult
import com.example.the_machine.operator.result.OperatorResultCode
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class OperatorResultTest {

  @Test
  fun testSuccessFactory() {
    val inputs = mapOf("param1" to "value1")
    val outputs = mapOf("result" to "success")

    val result = OperatorResult.success("TestOp", "1.0", inputs, outputs)

    assertTrue(result.ok)
    assertEquals(OperatorResultCode.SUCCESS, result.code)
    assertEquals("Operation completed successfully", result.message)
    assertEquals("TestOp", result.operatorName)
    assertEquals("1.0", result.operatorVersion)
    assertEquals(inputs, result.normalizedInputs)
    assertEquals(outputs, result.outputs)
    assertEquals(1, result.attempts)
  }

  @Test
  fun testSuccessFactoryWithCustomMessage() {
    val inputs = mapOf("param1" to "value1")
    val outputs = mapOf("result" to "success")

    val result = OperatorResult.success("TestOp", "1.0", "Custom success", inputs, outputs)

    assertTrue(result.ok)
    assertEquals(OperatorResultCode.SUCCESS, result.code)
    assertEquals("Custom success", result.message)
    assertEquals("TestOp", result.operatorName)
    assertEquals("1.0", result.operatorVersion)
  }

  @Test
  fun testFailureFactory() {
    val inputs = mapOf("param1" to "value1")

    val result = OperatorResult.failure(
      "TestOp", "1.0", OperatorResultCode.VALIDATION_ERROR,
      "Parameter validation failed", inputs
    )

    assertFalse(result.ok)
    assertEquals(OperatorResultCode.VALIDATION_ERROR, result.code)
    assertEquals("Parameter validation failed", result.message)
    assertEquals("TestOp", result.operatorName)
    assertEquals("1.0", result.operatorVersion)
    assertEquals(inputs, result.normalizedInputs)
    assertTrue(result.outputs.isEmpty())
  }

  @Test
  fun testValidationFailureFactory() {
    val result = OperatorResult.validationFailure(
      "TestOp", "1.0",
      "Required parameter 'name' is missing"
    )

    assertFalse(result.ok)
    assertEquals(OperatorResultCode.VALIDATION_ERROR, result.code)
    assertEquals("Required parameter 'name' is missing", result.message)
    assertEquals("TestOp", result.operatorName)
    assertEquals("1.0", result.operatorVersion)
  }

  @Test
  fun testWithAttempts() {
    val original = OperatorResult.success("TestOp", "1.0", emptyMap(), emptyMap())
    val updated = original.withAttempts(3)

    assertEquals(1, original.attempts)
    assertEquals(3, updated.attempts)

    // Ensure other fields are preserved
    assertEquals(original.ok, updated.ok)
    assertEquals(original.code, updated.code)
    assertEquals(original.operatorName, updated.operatorName)
  }

  @Test
  fun testWithMetrics() {
    val original = OperatorResult.success("TestOp", "1.0", emptyMap(), emptyMap())
    val additionalMetrics = mapOf("latency" to 150L, "retries" to 2)
    val updated = original.withMetrics(additionalMetrics)

    assertTrue(original.metrics.isEmpty())
    assertEquals(150L, updated.metrics["latency"])
    assertEquals(2L, updated.metrics["retries"])

    // Test merging with existing metrics
    val existingMetrics = mapOf("cpu" to 0.8)
    val withExisting = OperatorResult.success("TestOp", "1.0", emptyMap(), emptyMap())
      .withMetrics(existingMetrics)

    val merged = withExisting.withMetrics(additionalMetrics)
    assertEquals(0.8, merged.metrics["cpu"])
    assertEquals(150L, merged.metrics["latency"])
    assertEquals(2L, merged.metrics["retries"])
  }

  @Test
  fun testWithRemediationPatches() {
    val original = OperatorResult.success("TestOp", "1.0", emptyMap(), emptyMap())
    val patch = KotlinSerializationConfig.staticJson.parseToJsonElement(
      """
      {"op": "replace", "path": "/param1", "value": "newValue"}
    """
    )
    val patches = listOf<JsonElement>(patch)

    val updated = original.withRemediationPatches(patches)

    assertTrue(original.remediationPatches.isEmpty())
    assertEquals(1, updated.remediationPatches.size)
    assertEquals(
      "replace",
      updated.remediationPatches.first().jsonObject["op"]!!.jsonPrimitive.content
    )
  }

  @Test
  fun testToString() {
    val result = OperatorResult.success("TestOperator", "2.1", emptyMap(), emptyMap())
    val toString = result.toString()

    assertTrue(toString.contains("ok=true"))
    assertTrue(toString.contains("code=SUCCESS"))
    assertTrue(toString.contains("operatorName='TestOperator'"))
    assertTrue(toString.contains("operatorVersion='2.1'"))
    assertTrue(toString.contains("attempts=1"))
    assertTrue(toString.contains("message='Operation completed successfully'"))
  }

  @Test
  fun testNullHandling() {
    // Test with empty inputs and outputs (Kotlin doesn't accept null, but behavior should be same)
    val result = OperatorResult.success("TestOp", "1.0", emptyMap(), emptyMap())

    assertNotNull(result.normalizedInputs)
    assertNotNull(result.outputs)
    assertTrue(result.normalizedInputs.isEmpty())
    assertTrue(result.outputs.isEmpty())
  }

  @Test
  fun testCustomCodeViaFailureMethod() {
    val result = OperatorResult.failure(
      "CustomOp", "3.0", OperatorResultCode.EXECUTION_FAILED,
      "Custom operation completed", emptyMap()
    )

    assertFalse(result.ok) // failure method creates failed results
    assertEquals(OperatorResultCode.EXECUTION_FAILED, result.code)
    assertEquals("Custom operation completed", result.message)
    assertEquals("CustomOp", result.operatorName)
    assertEquals("3.0", result.operatorVersion)
    assertEquals(1, result.attempts) // default attempts
  }
}